src/derive_util.rs

```diff
@@ -62,3 +62,68 @@ macro_rules! union_has_padding {
         false $(|| core::mem::size_of::<$t>() != core::mem::size_of::<$ts>())*
     };
 }
+
+/// Implements `TryFromBytes` for a struct type by delegating to existing
+/// implementations for each of its fields, and optionally supports a custom
+/// validation method.
+///
+/// ```rust
+/// # use zerocopy::impl_try_from_bytes_for_struct;
+///
+/// #[repr(C)]
+/// struct Foo {
+///     a: u8,
+///     b: u16,
+/// }
+///
+/// impl_try_from_bytes_for_struct!(Foo { a: u8, b: u16 });
+///
+/// #[repr(transparent)]
+/// struct Bar(Foo);
+///
+/// impl Bar {
+///     fn is_valid(&self) -> bool {
+///         u16::from(self.0.a) < self.0.b
+///     }
+/// }
+///
+/// impl_try_from_bytes_for_struct!(Bar { 0: Foo } => is_valid);
+/// ```
+///
+/// # Safety
+///
+/// `$ty` must be a struct type, `$f` must list every field's name, and `$f_ty`
+/// must be the correct types for those fields.
+#[doc(hidden)] // `#[macro_export]` bypasses this module's `#[doc(hidden)]`.
+#[macro_export]
+macro_rules! impl_try_from_bytes_for_struct {
+    ($ty:ty { $($f:tt: $f_ty:ty),* } $(=> $validation_method:ident)?) => {
+        // SAFETY: The caller promises that all fields are listed with their
+        // correct types. We validate that every field is valid, which is the
+        // only requirement for the entire struct to be valid. Thus, we
+        // correctly implement `is_bit_valid` as required by the trait's safety
+        // documentation.
+        #[allow(unused_qualifications)]
+        unsafe impl zerocopy::TryFromBytes for $ty {
+            fn is_bit_valid(bytes: &zerocopy::MaybeValid<Self>) -> bool {
+                true $(&& {
+                    let f: &zerocopy::MaybeValid<$f_ty> = zerocopy::project!(&bytes.$f);
+                    zerocopy::TryFromBytes::is_bit_valid(f)
+                })*
+                $(&& {
+                    // SAFETY: We just validated that all of the struct's fields
+                    // are valid, which means that the struct itself is valid.
+                    // That is the only precondition of `assume_valid_ref`.
+                    let slf = unsafe { bytes.assume_valid_ref() };
+                    // TODO: What about interior mutability? One approach would
```

Comment:
For now, we just forbid interior mutability for `TryFromBytes`

Reply:
Updated the comment.

While we're on the subject, though, how do we actually ban interior mutability in `MaybeValid`? Currently, we implement `AsMaybeUninit` for all sized types and slices of sized types, which means that impls like the following exist automatically thanks to those blanket impls:

```rust
unsafe impl<T> AsMaybeUninit for UnsafeCell<T> {
    type MaybeUninit = MaybeUninit<UnsafeCell<T>>;
}
```

That, in turn, means that it's valid to write `MaybeValid<UnsafeCell<T>>`.

Maybe we could add safety preconditions to methods like `assume_valid_ref` that `T::MaybeUninit` doesn't contain any `UnsafeCell`s? Then, so long as we only implement `TryFromBytes` for types without `UnsafeCell`s, and we only ever emit calls to `assume_valid_ref` in such derived impls, then we should be okay.

Any other ideas? It'd be nice to not have this be so viral.

Reply:
> how do we actually ban interior mutability in MaybeValid

There's no reason to. The only place where shared mutability gets involved in `MaybeValid` is through the `TryFromBytes::is_bit_valid` method - that's the one that needs guarantees that there is only a single reference, either through `unsafe` or through banning `UnsafeCell` for `TryFromBytes` entirely.

> That, in turn, means that it's valid to write MaybeValid<UnsafeCell<T>>.

Yes, you can go from `&mut MaybeValid<T>` to `&mut T` on any type via a `.write()`. That's fine though - this doesn't involve shared mutability due to `&mut`.

> Maybe we could add safety preconditions to methods like assume_valid_ref that T::MaybeUninit doesn't contain any UnsafeCells

Again, I don't see why this is in _any_ way necessary.

> See https://github.com/rust-lang/unsafe-code-guidelines/issues/455 for more discussion.

That's essentially everything I told you offline - SB bans it, TB is fine. `MaybeUninit` does not interact with shared mutability, and neither should the `AsMaybeUninit` trait.

Reply:
Okay, I believe I've completely handled the interior mutability issue with [this update](https://github.com/google/zerocopy/compare/1ecc13acb232d7916a29ff64e9fa59951f7d55ab..6a4ff328652e1dc4fc7782bd79571c97295d3df1). Lmk if this looks sound to you?

For a bit more detail on my thinking (I think we're both on the same page now, but just to be sure), I had two concerns:
- Is it sound to have `&MaybeValid<T>` and `&T` at the same time? This is required in order to call custom validators.
  - This is now sound thanks to the requirement that `T` and `T::MaybeUninit` have `UnsafeCell`s at the same offsets.
- Is it sound to have `&MaybeValid<T>` and `&[u8]` at the same time? This is required in order to implement `try_from_ref` and friends since they take a `&[u8]` and convert it to a `&MaybeValid<T>` in order to call `is_bit_valid`.
  - This is sound so long as `T::MaybeUninit` doesn't contain any `UnsafeCell`s. Thanks to the `UnsafeCell`s-at-the-same-locations requirement, this is sound so long as `T` doesn't contain any `UnsafeCell`s. Since we only implement `FromZeroes` and `FromBytes` for `MaybeValid<T>` where `T: TryFromBytes` (and since `T: TryFromBytes` bans `UnsafeCell`s), we're okay. Given those impls, we only use safe code (namely, `Ref` and `FromBytes` methods) to convert `&[u8]` to `&MaybeValid<T>`.